backends/sycl/ceed-sycl-common.hpp

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2017-2022, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+#ifndef _ceed_sycl_common_hpp
+#define _ceed_sycl_common_hpp
+
+#include <ceed/backend.h>
+#include <ceed/jit-source/sycl/sycl-types.h>
+
+#include <sycl/sycl.hpp>
+#include <type_traits>
+
+#define CeedCallSycl(ceed, ...)                               \
+  do {                                                        \
+    try {                                                     \
+      __VA_ARGS__;                                            \
+    } catch (sycl::exception const &e) {                      \
+      return CeedError((ceed), CEED_ERROR_BACKEND, e.what()); \
+    }                                                         \
+  } while (0);
+
+using CeedBackendFunction = int (*)();
+
+template <typename R, class... Args>
+int CeedSetBackendFunctionCpp(Ceed ceed, const char *type, void *object, const char *fname, R (*f)(Args...)) {
+  static_assert(std::is_same_v<int, R>, "Ceed backend functions must return int");
+  // Kris: this is potentially undefined behavior by C++ standards
+  auto *bf = reinterpret_cast<CeedBackendFunction>(f);
+  return CeedSetBackendFunction(ceed, type, object, fname, bf);
+}
+
+typedef struct {
+  sycl::context sycl_context;
+  sycl::device  sycl_device;
+  sycl::queue   sycl_queue;
+} Ceed_Sycl;
+
+CEED_INTERN int CeedSyclGetResourceRoot(Ceed ceed, const char *resource, char **resource_root);
```

Comment:
Probably `CeedGetResourceRoot` should be a `ceed/backend.h` function in `interface/ceed.c`

Reply:
Maybe with the function signature `CeedGetResourceRoot(Ceed ceed, char **resource_root)` or `CeedGetResourceRoot(Ceed ceed, const char *deliminator, char **resource_root)`

Reply:
Ok if I leave that for a later PR?

Reply:
Sure, if we need to merge this today

Reply:
IDK about today, but soon. I'm guessing by your proposal that other backends would also need to implement a `CeedGetResourceRoot` method?

Reply:
I don't see this as a backend specific thing? That's why I suggested moving the entire function to ceed.c instead of making it only a sycl backend function. I don't see any special logic that only applies to sycl in this function, except maybe the specific delineator, which can be passed as an argument.

This is something other GPU backbends have discussed wanting

Reply:
It looks like all the GPU backends have their own version of `Ceed[ARCH]GetResourceRoot`, which is kind of why I suggested another PR (let this one be just SYCL, then we can merged all the GPU backends together in a separate change).

Reply:
Ah, I hadn't realized the idea had already invaded in this fractured fashion